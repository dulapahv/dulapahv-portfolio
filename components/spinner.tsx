import * as React from 'react';

import { cn } from '@/lib/utils';

export interface SpinnerProps extends React.HTMLAttributes<HTMLDivElement> {
  className?: string;
  variant?: 'default' | 'primary' | 'secondary';
  size?: number | 'default';
}

const Spinner = ({ className, size = 'default' }: SpinnerProps) => (
  <div
    role="status"
    aria-label="Loading"
    className={cn('[&>div]:bg-foreground size-5', className)}
    style={typeof size === 'number' ? { width: size, height: size } : undefined}
  >
    {Array.from({ length: 12 }).map((_, i) => (
      <div
        key={i}
        className="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%]
          origin-[center_190%] rounded-full opacity-[0.1] will-change-transform"
        style={{
          transform: `rotate(${i * 30}deg)`,
          animationDelay: `${(i * 0.083).toFixed(3)}s`,
        }}
        aria-hidden="true"
      />
    ))}
    <span className="sr-only">Loading...</span>
  </div>
);

export { Spinner };
