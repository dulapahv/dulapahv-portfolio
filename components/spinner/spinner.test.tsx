import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import { Spinner } from './spinner';

describe('Spinner', () => {
  it('should render spinner with loading status', () => {
    render(<Spinner />);

    const spinner = screen.getByRole('status');
    expect(spinner).toBeInTheDocument();
    expect(spinner).toHaveAttribute('aria-label', 'Loading');
  });

  it('should render 12 spinner bars', () => {
    const { container } = render(<Spinner />);

    const spinnerBars = container.querySelectorAll('[aria-hidden="true"]');
    expect(spinnerBars).toHaveLength(12);
  });

  it('should render with default size', () => {
    const { container } = render(<Spinner />);
    const spinner = container.querySelector('[role="status"]');

    expect(spinner).not.toHaveAttribute('style');
  });

  it('should render with custom numeric size', () => {
    const { container } = render(<Spinner size={40} />);
    const spinner = container.querySelector('[role="status"]');

    expect(spinner).toHaveAttribute('style', 'width: 40px; height: 40px;');
  });

  it('should apply custom className', () => {
    const { container } = render(<Spinner className="custom-class" />);
    const spinner = container.querySelector('[role="status"]');

    expect(spinner).toHaveClass('custom-class');
  });

  it('should have screen reader text', () => {
    render(<Spinner />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('should apply correct rotation to each bar', () => {
    const { container } = render(<Spinner />);
    const spinnerBars = container.querySelectorAll('[aria-hidden="true"]');

    // Check first bar rotation
    expect(spinnerBars[0]).toHaveStyle({ transform: 'rotate(0deg)' });
    // Check second bar rotation
    expect(spinnerBars[1]).toHaveStyle({ transform: 'rotate(30deg)' });
    // Check last bar rotation
    expect(spinnerBars[11]).toHaveStyle({ transform: 'rotate(330deg)' });
  });

  it('should match snapshot', () => {
    const { container } = render(<Spinner />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with custom size', () => {
    const { container } = render(<Spinner size={50} />);
    expect(container).toMatchSnapshot();
  });
});
