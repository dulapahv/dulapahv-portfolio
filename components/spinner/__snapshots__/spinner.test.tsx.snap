// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spinner > should match snapshot 1`] = `
<div>
  <div
    aria-label="Loading"
    class="[&>div]:bg-foreground size-5"
    role="status"
  >
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(0deg); animation-delay: 0.000s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(30deg); animation-delay: 0.083s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(60deg); animation-delay: 0.166s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(90deg); animation-delay: 0.249s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(120deg); animation-delay: 0.332s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(150deg); animation-delay: 0.415s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(180deg); animation-delay: 0.498s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(210deg); animation-delay: 0.581s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(240deg); animation-delay: 0.664s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(270deg); animation-delay: 0.747s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(300deg); animation-delay: 0.830s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(330deg); animation-delay: 0.913s;"
    />
    <span
      class="sr-only"
    >
      Loading...
    </span>
  </div>
</div>
`;

exports[`Spinner > should match snapshot with custom size 1`] = `
<div>
  <div
    aria-label="Loading"
    class="[&>div]:bg-foreground size-5"
    role="status"
    style="width: 50px; height: 50px;"
  >
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(0deg); animation-delay: 0.000s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(30deg); animation-delay: 0.083s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(60deg); animation-delay: 0.166s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(90deg); animation-delay: 0.249s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(120deg); animation-delay: 0.332s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(150deg); animation-delay: 0.415s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(180deg); animation-delay: 0.498s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(210deg); animation-delay: 0.581s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(240deg); animation-delay: 0.664s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(270deg); animation-delay: 0.747s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(300deg); animation-delay: 0.830s;"
    />
    <div
      aria-hidden="true"
      class="animate-spinner absolute top-[4.4%] left-[46.5%] h-[24%] w-[7%] origin-[center_190%] rounded-full
            opacity-[0.1] will-change-transform"
      style="transform: rotate(330deg); animation-delay: 0.913s;"
    />
    <span
      class="sr-only"
    >
      Loading...
    </span>
  </div>
</div>
`;
